### PR TITLE
Optimise language service during configuration switches

### LINKED ISSUE #7615: Build inside VS may take forever due to a deadlock inside NET Core project
## Related Items - 
[AB#1439339](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1439339)
[AB#1439046](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1439046)

## Steps to Reproduce

1. Load a mid size solution into VS (like CPS)
2. Build the solution, works well
3. Switch solution configuration (debug -> release or reverse)
4. Build the solution immediately

## Expected Behavior
Build the solution 

## Actual Behavior
Build may hang, and will never finish.  It was reported earlier, but no clear repo.

I ran into this and captured a dump & can share if necessary.

Basically, build tasks are blocked in AppendFlushLoggersTasks

```
1f5490a7e08 <8> Microsoft.VisualStudio.ProjectSystem.Build.BuildProject+<BuildCoreAsync>d__100 @ 7fff3741d410
..1f5490a8160 <1> Microsoft.VisualStudio.ProjectSystem.Build.BuildProject+<BuildWithSolutionAsync>d__94 @ 7fff37992bb0
..1f5490a74c0 <5> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.BuildableProjectConfig+<BuildAsync>d__63 @ 7fff7f57a580

1f549300df8 <8> Microsoft.VisualStudio.ProjectSystem.Build.BuildProject+<BuildCoreAsync>d__100 @ 7fff3741d410
..1f549301150 <1> Microsoft.VisualStudio.ProjectSystem.Build.BuildProject+<BuildWithSolutionAsync>d__94 @ 7fff37992bb0
..1f5492f69f0 <5> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.BuildableProjectConfig+<BuildAsync>d__63 @ 7fff7f57a580

1f5491a8918 <0> Microsoft.VisualStudio.ProjectSystem.Build.BuildUtilities+<>c+<<AppendFlushLoggersTask>b__16_0>d @ 7fff388c9320
1f54930f080 <0> Microsoft.VisualStudio.ProjectSystem.Build.BuildUtilities+<>c+<<AppendFlushLoggersTask>b__16_0>d @ 7fff388c9320
```

Further tracking down, they were blocked by those tasks during error list initialization.

```
1f4b5929fe8 <0> Microsoft.VisualStudio.ProjectSystem.AbstractMultiLifetimeComponent<Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+WorkspaceProjectContextHostInstance>+<WaitForLoadedAsync>d__3 @ 7fff367540e0
..1f4b592a248 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+<OpenContextForWriteAsync>d__14 @ 7fff36775cc0
..1f4b592a498 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.ActiveWorkspaceProjectContextHost+<OpenContextForWriteAsync>d__5 @ 7fff36774bc0
..1f4b592a6e0 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<>c+<<ClearAllAsync>b__28_0>d @ 7fff7db718e0
..1f4b592bf78 <0> Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools+<ForEachAggregateExceptionsAsync>d__65<System.Lazy<Microsoft.VisualStudio.ProjectSystem.VS.Build.IVsErrorListProvider, Microsoft.VisualStudio.ProjectSystem.IOrderPrecedenceMetadataView>> @ 7fff38c05fc0
..1f4b592c1d0 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<ClearAllAsync>d__28 @ 7fff7db71570
..1f4b592c410 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<ProcessQueueAsync>d__31 @ 7fff79f87770
..1f4b592c6f8 <0> Microsoft.VisualStudio.ProjectSystem.TransformBlockSlim<Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.Build.ProjectBuildMessages>, System.Collections.Immutable.IImmutableDictionary<Microsoft.VisualStudio.ProjectSystem.NamedIdentity, System.IComparable>>+TransformBlockSlimAsync+<ProcessInputAsync>d__3 @ 7fff38877b20
..1f4b592c940 <0> Microsoft.VisualStudio.ProjectSystem.DataReceivingBlockSlim<Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.Build.ProjectBuildMessages>>+<ProcessInputQueueAsync>d__5 @ 7fff388771d0

1f4b819d430 <0> Microsoft.VisualStudio.ProjectSystem.AbstractMultiLifetimeComponent<Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+WorkspaceProjectContextHostInstance>+<WaitForLoadedAsync>d__3 @ 7fff367540e0
..1f4b819dbf8 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+<OpenContextForWriteAsync>d__14 @ 7fff36775cc0
..1f4b819de48 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.ActiveWorkspaceProjectContextHost+<OpenContextForWriteAsync>d__5 @ 7fff36774bc0
..1f4b819e090 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<>c+<<ClearAllAsync>b__28_0>d @ 7fff7db718e0
..1f4b819e2d0 <0> Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools+<ForEachAggregateExceptionsAsync>d__65<System.Lazy<Microsoft.VisualStudio.ProjectSystem.VS.Build.IVsErrorListProvider, Microsoft.VisualStudio.ProjectSystem.IOrderPrecedenceMetadataView>> @ 7fff38c05fc0
..1f4b819e528 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<ClearAllAsync>d__28 @ 7fff7db71570
..1f4b819e768 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<ProcessQueueAsync>d__31 @ 7fff79f87770
..1f4b819ea50 <0> Microsoft.VisualStudio.ProjectSystem.TransformBlockSlim<Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.Build.ProjectBuildMessages>, System.Collections.Immutable.IImmutableDictionary<Microsoft.VisualStudio.ProjectSystem.NamedIdentity, System.IComparable>>+TransformBlockSlimAsync+<ProcessInputAsync>d__3 @ 7fff38877b20
..1f4b819ec98 <0> Microsoft.VisualStudio.ProjectSystem.DataReceivingBlockSlim<Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.Build.ProjectBuildMessages>>+<ProcessInputQueueAsync>d__5 @ 7fff388771d0

1f4b8602d68 <0> Microsoft.VisualStudio.ProjectSystem.AbstractMultiLifetimeComponent<Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+WorkspaceProjectContextHostInstance>+<WaitForLoadedAsync>d__3 @ 7fff367540e0
..1f4b8602fc8 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+<OpenContextForWriteAsync>d__14 @ 7fff36775cc0
..1f4b8603218 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.ActiveWorkspaceProjectContextHost+<OpenContextForWriteAsync>d__5 @ 7fff36774bc0
..1f4b8603460 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<>c+<<ClearAllAsync>b__28_0>d @ 7fff7db718e0
..1f4b86036a0 <0> Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools+<ForEachAggregateExceptionsAsync>d__65<System.Lazy<Microsoft.VisualStudio.ProjectSystem.VS.Build.IVsErrorListProvider, Microsoft.VisualStudio.ProjectSystem.IOrderPrecedenceMetadataView>> @ 7fff38c05fc0
..1f4b86038f8 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<ClearAllAsync>d__28 @ 7fff7db71570
..1f4b8603b38 <0> Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Build.VsProjectBuildErrorListPublishService+<ProcessQueueAsync>d__31 @ 7fff79f87770
..1f4b8603e20 <0> Microsoft.VisualStudio.ProjectSystem.TransformBlockSlim<Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.Build.ProjectBuildMessages>, System.Collections.Immutable.IImmutableDictionary<Microsoft.VisualStudio.ProjectSystem.NamedIdentity, System.IComparable>>+TransformBlockSlimAsync+<ProcessInputAsync>d__3 @ 7fff38877b20
..1f4b8604068 <0> Microsoft.VisualStudio.ProjectSystem.DataReceivingBlockSlim<Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.Build.ProjectBuildMessages>>+<ProcessInputQueueAsync>d__5 @ 7fff388771d0
```

Here we can see the code has a race condition, the ActiveWorkspaceProjectContextHost uses IActiveConfiguredValue<IWorkspaceProjectContextHost?> _activeHost to access WorkspaceProjectContext. During the configuration switching, this logic may get the old WorkspaceProjectContext, because it is deactivated, WaitForLoadedAsync will hang forever to wait it to be reinitialized. When it hangs, however, the _activeHost now gets the right configuration.  It looks like the code expects ActiveProjectConfigurationChangedException, however, in its code path, it will block waits a TaskCompletionSource inside a configured project, and nothing will stop it and throws this Exception.

```
1f4b5929fe8 <0> Microsoft.VisualStudio.ProjectSystem.AbstractMultiLifetimeComponent<Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+WorkspaceProjectContextHostInstance>+<WaitForLoadedAsync>d__3 @ 7fff367540e0
..1f4b592a248 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.WorkspaceProjectContextHost+<OpenContextForWriteAsync>d__14 @ 7fff36775cc0
..1f4b592a498 <0> Microsoft.VisualStudio.ProjectSystem.LanguageServices.ActiveWorkspaceProjectContextHost+<OpenContextForWriteAsync>d__5 @ 7fff36774bc0
```

### PR DESCRIPTION
**Pull request by drewnoakes**
Addresses half of #7541 (the switching half, not the reload half).
Fixes #7615

This change alters how the language service constructs and manages Roslyn workspace objects.

Previously, a workspace would be created per `ConfiguredProject`. With this change, one is created per "slice".

The easiest way to explain slices is with an example. Consider the following configurations:

- `Debug|AnyCPU|net48`
- `Debug|AnyCPU|net60`
- `Release|AnyCPU|net48`
- `Release|AnyCPU|net60`

If `Debug|AnyCPU` is active in the IDE, then the first two configurations are active. If the user switches to `Release`, those two are deactivated, and the last two become active.

When considering slices, the configuration (eg. `Debug`/`Release`) and platform (eg. `AnyCPU`) are ignored. Only additional (implicit) dimensions are considered (eg. `TargetFramework`).

CPS provides a new facility (via `IActiveConfigurationGroupSubscriptionService`) that allows subscribing to data for (in this example) `net48` and `net60` slices of the project. If the user switches configuration, the same subscription is used, however any difference in the data between configurations is received (via Dataflow) as a delta.

Note that for single-targeting projects (i.e. a single target framework) there is only one slice for the project, and it still updates as described.

Given we now receive a delta when the configuration changes, we are able to pass that delta on to Roslyn. Roslyn will be able to optimise this process to avoid discarding all work done for one configuration, and starting fresh with another configuration when a switch occurs.

We don't have MEF scopes for slices, and slices do not belong to a specific configuration, so MEF components that work with slices needed to be moved from `ConfiguredProject` scope to `UnconfiguredProject` scope.

This change also reduces the number of types/interfaces/concepts involved in creating, updating and generally working with workspaces.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8321)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+///     To interrogate the active workspace as part of various IDE features, including acquiring a read or write lock.
+///   </item>
+/// </list>
+/// </remarks>
+[Export(typeof(IWorkspaceWriter))]
+[Export(ExportContractNames.Scopes.UnconfiguredProject, typeof(IProjectDynamicLoadComponent))]
+[AppliesTo(ProjectCapability.DotNetLanguageService)]
+internal sealed class LanguageServiceHost : OnceInitializedOnceDisposedUnderLockAsync, IProjectDynamicLoadComponent, IWorkspaceWriter
+{
+    // TODO don't activate in if _vsShellServices.Value.IsInCommandLineMode
```
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs:37]: What is this in reference to?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+                    _projectGuid,
+                    hostObject,
+                    binOutputPath,
+                    assemblyName,
+                    cancellationToken);
+
+                _disposableBag.Add(_context);
+
+                // Update additional properties within a batch to avoid thread pool starvation.
+                // https://github.com/dotnet/project-system/issues/8027
```
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs:221]: Looks like this issue was closed.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+        void ApplyProjectBuild(
+            IProjectVersionedValue<BuildUpdate> update,
+            ContextState state,
+            CancellationToken cancellationToken)
+        {
+            IProjectChangeDescription projectChange = update.Value.BuildRuleUpdate.ProjectChanges[ProjectBuildRuleName];
+
+            // There should always be some change to publish, as we have already called BeginBatch by this point
+            // TODO understand why the CLA snapshot's changed state differs from the project update, as they are supposed to travel together in sync
+            //Assumes.True(projectChange.Difference.AnyChanges && update.Value.CommandLineArgumentsSnapshot.IsChanged);
```
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs:367]: nit: Commented code.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceFactory.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            ProjectDataSources.JoinUpstreamDataSources(joinableTaskFactory, _projectService.Services.FaultHandler, source.ProjectRuleSource, source.SourceItemsRuleSource)
+        });
+
+        #endregion
+
+        #region Build data subscriptions
+
+        // A block that provides the command line arguments for a configured project.
+        // We will link this to the slice's active configured project.
+        // NOTE this will go away once CPS exposes a way to access snapshot data while preserving order (being worked on for 17.4)
```
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceFactory.cs:118]: Link for tracking?

**Review by MiYanni — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs:37]: https://github.com/dotnet/project-system/issues/3832

I'll include that in the code directly too.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs (hunk 2 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs:221]: Yeah. This comment was added in the PR that closed it: https://github.com/dotnet/project-system/pull/8031/files#diff-1844e8887b42f22e08839aadff0e0e290a54b9e62bfc316f7c3fdb851d502cbcR82-R83

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs (hunk 3 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs:367]: The TODO relates to that. I plan to revisit the use of CommandLineArguments in a future PR that'll sort this out.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceFactory.cs (hunk 4 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceFactory.cs:118]: CPS has already implemented this. I have a local branch that'll address this. I don't want to make this PR any more complex though.

File: eng/imports/Packages.targets [hunk 5]
```diff
@@ -89,12 +88,14 @@
...
 
     <!-- CPS -->
+    <!-- Find versions at https://dev.azure.com/azure-public/vside/_artifacts/feed/vs-impl -->
     <PackageReference Update="Microsoft.VisualStudio.ProjectSystem"                                   Version="17.4.5-pre" />
     <PackageReference Update="Microsoft.VisualStudio.ProjectSystem.Analyzers"                         Version="17.4.5-pre" />
     <PackageReference Update="Microsoft.VisualStudio.ProjectSystem.Query"                             Version="17.4.5-pre" />
     <PackageReference Update="Microsoft.VisualStudio.ProjectSystem.SDK.Tools"                         Version="17.4.5-pre" />
 
     <!-- Roslyn -->
+    <PackageReference Update="Microsoft.Net.Compilers.Toolset"                                        Version="4.3.0-2.final" />
```
Comment (MiYanni) [inline eng/imports/Packages.targets:98]: Thanks for moving this to the appropriate section! <3

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs (hunk 3 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs:403]: The PR that'll remove the `CommandLineArgumentsSnapshot` is #8337.

File: docs/repo/language-service.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+    E1 --> E2
+    B1 --> B2
+
+    E2 --> A
+    B2 --> A
+
+    A --> OnWorkspaceUpdateAsync
+
+  end
```
Comment (MiYanni) [inline docs/repo/language-service.md:94]: To make the lines less crissy crossy, I've modified the connections order.
```mermaid
graph LR;
  subgraph Slice
    ActiveConfiguredProjectSource
    ProjectRuleSource
    SourceItemsRuleSource
    ProjectBuildRuleSource
  end
  subgraph Workspace
    E1(Evaluation sync link)
    E2(Evaluation transform)
    B1(Build sync link)
    B2(Build transform)
    A(Action);

    ProjectRuleSource --> E1
    SourceItemsRuleSource --> E1
    ActiveConfiguredProjectSource --> E1

    ActiveConfiguredProjectSource --> B1
    ProjectBuildRuleSource --> B1

    E1 --> E2
    B1 --> B2

    E2 --> A
    B2 --> A

    A --> OnWorkspaceUpdateAsync

  end
```

```suggestion
graph LR;
  subgraph Slice
    ActiveConfiguredProjectSource
    ProjectRuleSource
    SourceItemsRuleSource
    ProjectBuildRuleSource
  end
  subgraph Workspace
    E1(Evaluation sync link)
    E2(Evaluation transform)
    B1(Build sync link)
    B2(Build transform)
    A(Action);

    ProjectRuleSource --> E1
    SourceItemsRuleSource --> E1
    ActiveConfiguredProjectSource --> E1

    ActiveConfiguredProjectSource --> B1
    ProjectBuildRuleSource --> B1

    E1 --> E2
    B1 --> B2

    E2 --> A
    B2 --> A

    A --> OnWorkspaceUpdateAsync

  end
```

Comment (MiYanni) [inline docs/repo/language-service.md:94]: To make the lines less crissy crossy, I've modified the connections order.
```mermaid
graph LR;
  subgraph Slice
    ActiveConfiguredProjectSource
    ProjectRuleSource
    SourceItemsRuleSource
    ProjectBuildRuleSource
  end
  subgraph Workspace
    E1(Evaluation sync link)
    E2(Evaluation transform)
    B1(Build sync link)
    B2(Build transform)
    A(Action);

    ProjectRuleSource --> E1
    SourceItemsRuleSource --> E1
    ActiveConfiguredProjectSource --> E1

    ActiveConfiguredProjectSource --> B1
    ProjectBuildRuleSource --> B1

    E1 --> E2
    B1 --> B2

    E2 --> A
    B2 --> A

    A --> OnWorkspaceUpdateAsync

  end
```

```suggestion
graph LR;
  subgraph Slice
    ActiveConfiguredProjectSource
    ProjectRuleSource
    SourceItemsRuleSource
    ProjectBuildRuleSource
  end
  subgraph Workspace
    E1(Evaluation sync link)
    E2(Evaluation transform)
    B1(Build sync link)
    B2(Build transform)
    A(Action);

    ProjectRuleSource --> E1
    SourceItemsRuleSource --> E1
    ActiveConfiguredProjectSource --> E1

    ActiveConfiguredProjectSource --> B1
    ProjectBuildRuleSource --> B1

    E1 --> E2
    B1 --> B2

    E2 --> A
    B2 --> A

    A --> OnWorkspaceUpdateAsync

  end
```

**Review by jasonmalinowski — APPROVED** [review]
So I haven't looked at the code, but testing a bunch of debug/release switches and looking under a debugger it appears to be doing what I'd expect from a language service perspective. So that's good!

**Pull request merged**